### ISSUE TITLE
Better explaination with asyncio

### ISSUE THREAD

**Issue by skrew** [etataurov/asyncio-apns#6]
Hi,

Your example is about sending a simple push, but in major case we need to send thousand of pushs.

I have a problem with it. I use a Semaphore for sending 50 pushs in concurrency (without it it's so slow ! about 2 pushs / sec) but i can't catch a bad token.

EG: I have a test push server, and with 188th push, i send back a 400: BadToken error.

Client side, i got the error but not for the good push/token (190th, 192th, even 189th but not the good one (188)

I have read there are something with "futures" but don't see many good examples on web about handling / playing with that.

My current testing code:

```
async def send_push4(apns, message, token):
    with await mysem:
        try:
            await apns.send_message(message, token)
        except Exception as APNsError:
            print("Error %s for %s" %(APNsError.status, token))


loop = asyncio.get_event_loop()
sem = asyncio.Semaphore(50)
for i in range(1,200):
        tasks.append(asyncio.async(send_push4(apns, message, token)))
loop.run_until_complete(asyncio.tasks)
```

Thanks for helping.

**Comment by etataurov** [comment]
@skrew Thanks for trying this lib

Am I getting your problem right?
188th request receives response with error, but you can't see it. 
189-200 requests should be successful, but you are seeing errors

If it is this case - it might be a bug, I'll investigate it

I can see the idea in your code, but it seems unworking. 
Could you please provide working one to be completely sure about it

**Comment by skrew** [comment]
Hi @etataurov,

No, i send (for testing) 200 pushs. 
For push 188 i make a BadToken one (i append -188 in token)
What i expect is to have, in exception, the bad token (with -188) (for flagging "bad token" in my database)
But actually i got others tokens in the except. 

As i understand, when the except code are executed, the variable `token` are already assigned to other token (because of async code)

I look for a way to be sure to have the good token in the except block, but i think it's not the good way to do that.

As i understand, there are something to do with "future" but i don't know how. 

As your lib are async compliant, is there a way to get the token who cause the Except code ? (bad token or any other reason) ?

**Comment by etataurov** [comment]
> As i understand, when the except code are executed, the variable token are already assigned to other token (because of async code)

In this code it is not. Token in except block will be the same as in send_message call. Because `token` is a string and passed into function by value, therefore it can't change

``` python
async def send_push4(apns, message, token):
    with await mysem:
        try:
            await apns.send_message(message, token)
        except Exception as APNsError:
            print("Error %s for %s" %(APNsError.status, token))
```

Are you sure you have such problem with this code?

> As your lib are async compliant, is there a way to get the token who cause the Except code ? (bad token or any other reason) ?

Unfortunately no, but it is useful and I will probably add it

**Comment by skrew** [comment]
Ok so it's maybe a problem with my test server... i will look and tell you.

Thanks.

**Comment by skrew** [comment]
Back... Just tested with a Go APNS client. My "mock" server are ok, and i got error for the good token each time. 

Whats can cause problem with your lib and python 3.5 (using asyncio) ? Any idea ?

**Comment by etataurov** [comment]
I don't know what the problem now
Could you please provide exact code that you are using and which has this problem.
I'll try to figure this out.

Thanks

**Comment by skrew** [comment]
I'm not at home ATM. Do you have GO installed ? I will send you client and server later

**Comment by skrew** [comment]
Hi,
Test script: https://dl.dropboxusercontent.com/u/71060639/scripts.tar.gz

_Note: you have to use asyncio-apns included (it's automatic in script) because it's a patched one for custom host / port and allow unsigned certificates. I've added support for `apns-priority` and `apns-topic` too..._

**How to use:**
- Run the server with `go run go-apns-server.go`
- Launch (4-5 times) `makePush2.py` (python 3.5 required), you can see the token don't always return the good one (188 are marked as badtoken, look in `go-apns-server.go` code).
- Launch `go run client.go`: you can see the token reported as bad is always the 188th, as expected.

Thanks.

**Comment by etataurov** [comment]
Hi, Thanks for providing these files!
I'll take a look today

**Comment by etataurov** [comment]
@skrew 
Hi, I think I've found and fixed bug. 
There was an error due to my initial misunderstanding of HTTP/2 streams. 
Thanks for helping me find it
Please try revision @6e40af3
It would be great to remove HTTP/2 client from this library and keep it very small.

I will appreciate if you make PR with your changes, but if you're busy I can merge them myself.

**Comment by skrew** [comment]
Hi @etataurov 
PR submitted.

I have to test your changes in real situation, but look good on mock server. You can confirm i'm doing right with Semaphore / await ? 

Thanks for helping ;)

**Comment by skrew** [comment]
Ok my exemple don't seem to works... i got the error "h2.exceptions.TooManyStreamsError: Max outbound streams is 250, 250 open" when i try with 40.000 pushs... I don't know what's the problem with the semaphore, i think i will just use golang for the push part, i had spend too many time on python.

**Comment by etataurov** [comment]
@skrew 
I think the problem is that stream is closed later than send_message future is resolved.
I opened #9 to make batch sends more user-friendly

I'm sorry that this library can't help you now.
But really thank you for testing it with cases I don't use

**Comment by skrew** [comment]
@etataurov Just look with my test code. Change range from 200 to 40.000, you will see there are errors before limits hit.

`Cannot send 34 bytes, flow control window is 17.`

Somewhere here, connection are not closed (i think).

I need to send a batch of about 200.000 pushs...

**Comment by etataurov** [comment]
TooManyStreamsError fixed in #9 
for window overflow I created new issue #10
